openhands/controller/agent_controller.py

```diff
@@ -420,13 +392,8 @@ async def _step(self) -> None:
             await asyncio.sleep(1)
             return
 
-        # check if agent got stuck before taking any action
         if self._is_stuck():
-            # This need to go BEFORE report_error to sync metrics
-            self.event_stream.add_event(
-                FatalErrorObservation('Agent got stuck in a loop'),
-                EventSource.ENVIRONMENT,
-            )
+            await self._react_to_exception(RuntimeError('Agent got stuck in a loop'))
```

Comment:
I think moving this check at the start of the `_step` might mean that you can never restore a session that ended in a loop... I didn't try it on the latest version, but I did on others; and well, it either works to stop or it doesn't...

This is not the same if it's at the end; then it stops, and then, upon session restore, the LLM has another chance this time. (I'll test the exact behavior, it might be a little non-obvious if it stops in `main.py`)

On a side note, I stopped merging stuff in this branch, since I noticed it's probably making it harder for you. Sorry about that. On this call, I did think that when this branch works fine, it _can_ be at the end again. I could be wrong?

Reply:
Just did a test--session recovery seems to work just fine if it gets stuck in a loop